src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/BlockReflectionTypeMapNode.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+
+using ILCompiler.Metadata;
+
+using Internal.TypeSystem;
+using Internal.NativeFormat;
+using Internal.Text;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents a map between EETypes and metadata records within the <see cref="MetadataNode"/>.
+    /// </summary>
+    internal sealed class BlockReflectionTypeMapNode : ObjectNode, ISymbolNode
+    {
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        private ExternalReferencesTableNode _externalReferences;
+
+        public BlockReflectionTypeMapNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__block_reflection_type_map_End", true);
+            _externalReferences = externalReferences;
+        }
+
+        public ISymbolNode EndSymbol => _endSymbol;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__block_reflection_type_map");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.DataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName() => this.GetMangledName();
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+
+            var writer = new NativeWriter();
+            var reflectionBlockTypeMapHashTable = new VertexHashtable();
+
+            Section hashTableSection = writer.NewSection();
+            hashTableSection.Place(reflectionBlockTypeMapHashTable);
+            IMetadataPolicy metadataPolicy = factory.MetadataManager.GetMetadataPolicy();
+
+            foreach (var type in factory.MetadataManager.GetTypesWithEETypes())
+            {
+                if (!type.IsDefType)
+                    continue;
+
+                if (!type.IsTypeDefinition)
+                    continue;
+
+                var mdType = type as MetadataType;
+                if (mdType == null)
+                    continue;
+
+                if (!metadataPolicy.IsBlocked(mdType))
```

Comment:
No, I thought it was a fairly nice scheme, but I'm happy enough to make it more specific.